raft-tech/TANF-app
# Fix parser/preparser validation of empty strings

## Summary of Changes
Due to [the way python slices strings given an out-of-bounds range](https://www.tutorialspoint.com/Why-substring-slicing-index-out-of-range-works-in-Python#:~:text=(substring)-,Slicing,-with%20an%20out), the parser was creating blank-string-filled records when a multi-record was not space-filled. Once section 1 validation was introduced in #2518, uncaught exceptions are thrown during parsing 

E.g.
`T320210400028221R0112014122888175617622222112204398100000000`
Creates two records, a properly filled t3 record, and a t3 consisting of `''` values for all char fields (which subsequently fails validation and ends the parser task for that submission)

whereas
`T320210400028221R0112014122888175617622222112204398100000000                                         `
Creates a single, properly filled t3 record with no errors.

I also moved around some functions and classes to resolve circular imports.

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-backend && docker-compose up
cd tdrs-frontend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Upload a file with the row included above, with and without the trailing space-fill
3. Ensure the resulting record, status, and error report are the same for both submissions.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
